qldpc/stim/util.py

```diff
@@ -0,0 +1,142 @@
+from __future__ import annotations
+
+import dataclasses
+from collections.abc import Callable
+
+import numpy as np
+import numpy.typing as npt
+import scipy.sparse
+import stim
+
+from qldpc.objects import Pauli, PauliXZ
+
+
+@dataclasses.dataclass(frozen=True)
+class CircuitLevelError:
+    dets: tuple[stim.DemTargetWithCoords, ...]
+    obs: tuple[stim.DemTarget, ...]
+    basis: PauliXZ
+
+
+@dataclasses.dataclass
+class CheckMatrices:
+    check_map: scipy.sparse.csc_matrix
+    check_matrix: scipy.sparse.csc_matrix
+    obs_matrix: scipy.sparse.csc_matrix
+    priors: npt.NDArray[np.float64]
+
+
+def _det_basis_coord(det: stim.DemTargetWithCoords) -> PauliXZ:
+    """
+    Returns the basis of the detector based on the 1st coordinate (1 == X, 2 == Z)
+    """
+    if det.coords[0] == 1:
+        return Pauli.X
+    elif det.coords[0] == 2:
+        return Pauli.Z
+    else:
+        raise ValueError(f"Invalid basis: {det.coords[0]} (must be 1 or 2)")
+
+
+def _prior_dict_to_matrices(
+    prior_dict: dict[CircuitLevelError, float], num_detectors: int, num_obs: int
+) -> CheckMatrices:
+    det_list: list[stim.DemTarget] = []
+    det_map: dict[stim.DemTarget, int] = {}
+    det_row_idx: list[int] = []
+    det_col_idx: list[int] = []
+
+    obs_list: list[int] = list(range(num_obs))
+    obs_row_idx: list[int] = []
+    obs_col_idx: list[int] = []
+
+    priors: list[float] = []
+
+    for i, (c_err, prior) in enumerate(prior_dict.items()):
+        priors.append(prior)
+
+        for det in c_err.dets:
+            det_val = det.dem_target.val
+            if det not in det_list:
+                det_map[det_val] = len(det_list)
+                det_list += [det]
+            det_row_idx += [det_map[det_val]]
+            det_col_idx += [i]
+
+        for obs in c_err.obs:
+            obs_row_idx += [obs.val]
+            obs_col_idx += [i]
+
+    # Resulting check matrix may have fewer dets than original
+    check_map = scipy.sparse.csc_matrix(
+        (np.ones(len(det_map)), (list(det_map.values()), list(det_map.keys()))),
+        shape=(len(det_list), num_detectors),
+    )
```

Comment:
I'm a bit confused about what the `check_map` is, and why it appears to be used as the effective check matrix (instead of `check_matrix`) in `CompiledSinterDecoder`.

Reply:
Oh -- is this just a remapping of detector coordinates?  If so, let's just merge it into the definitions of `check_matrix` and `obs_matrix`.

Reply:
Nevermind, I understand why this exists.  I'll try to explain carefully in code comments.  Resolving.